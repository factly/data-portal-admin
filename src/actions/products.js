import axios from '../utils/axios';
import {
  baseUrl,
  ADD_PRODUCTS_LIST_REQUEST,
  SET_PRODUCTS_LIST_CURRENT_PAGE,
  LOADING_PRODUCTS,
  LOAD_PRODUCTS_SUCCESS,
  SET_PRODUCTS_LIST_TOTAL,
  LOAD_PRODUCTS_FAILURE,
  CREATING_PRODUCT,
  CREATE_PRODUCT_SUCCESS,
  CREATE_PRODUCT_FAILURE,
  LOADING_PRODUCT_DETAILS,
  GET_PRODUCT_DETAILS_SUCCESS,
  GET_PRODUCT_DETAILS_FAILURE,
  UPDATING_PRODUCT,
  UPDATE_PRODUCT_SUCCESS,
  UPDATE_PRODUCT_FAILURE,
  DELETING_PRODUCT,
  DELETE_PRODUCT_SUCCESS,
  DELETE_PRODUCT_FAILURE,
} from '../constants/products';
import { loadCategoriesSuccess } from './categories';
import { loadCurrenciesSuccess } from './currencies';
import { loadProductTypesSuccess } from './product_types';
import { loadTagsSuccess } from './tags';
import { getIds, getValues, deleteKeys, buildObjectOfItems } from '../utils/objects';

export const loadProducts = (page, limit) => {
  return async (dispatch, getState) => {
    let url = baseUrl;
    if (page && limit) {
      url = `${url}?page=${page}&limit=${limit}`;
    }

    const {
      products: { req },
    } = getState();

    let found = false;
    let ids;
    for (let item of req) {
      if (item.page === page && item.limit === limit) {
        ids = [...item.ids];
        found = true;
      }
    }

    if (found) {
      dispatch(setListCurrentPage(ids));
      return;
    }

    dispatch(loadingProducts());

    const response = await axios({
      url: url,
      method: 'get',
    }).catch((error) => {
      dispatch(loadProductsFailure(error.message));
    });

    if (response) {
      const { nodes, total } = response.data;
      const currentPageIds = getIds(nodes);
      const req = { page: page, limit: limit, ids: currentPageIds };

      const categories = getValues(nodes, 'categories');
      dispatch(loadCategoriesSuccess(categories));

      const currencies = getValues(nodes, 'currency');
      dispatch(loadCurrenciesSuccess(currencies));

      const productTypes = getValues(nodes, 'product_type');
      dispatch(loadProductTypesSuccess(productTypes));

      const tags = getValues(nodes, 'tags');
      dispatch(loadTagsSuccess(tags));

      nodes.forEach((product) => {
        product.categories = getIds(product.categories);
        product.tags = getIds(product.tags);
      });

      dispatch(addListRequest(req));
      dispatch(loadProductsSuccess(nodes));
      dispatch(setListCurrentPage(currentPageIds));
      dispatch(setProductsListTotal(total));
    }
  };
};

export const createProduct = (data) => {
  return async (dispatch, getState) => {
    dispatch(creatingProduct());

    const response = await axios({
      url: baseUrl,
      method: 'post',
      data: data,
    }).catch((error) => {
      dispatch(createProductFailure(error.message));
    });

    if (response) {
      dispatch(createProductSuccess(response.data));
    }
  };
};

export const getProductDetails = (id) => {
  return async (dispatch, getState) => {
    let url = `${baseUrl}/${id}`;

    dispatch(loadingProductDetails());

    const response = await axios({
      url: url,
      method: 'get',
    }).catch((error) => {
      dispatch(getProductDetailsFailure(error.message));
    });

    if (response) {
      dispatch(getProductDetailsSuccess(response.data));
    }
  };
};

export const updateProduct = (id, data) => {
  return async (dispatch, getState) => {
    let url = `${baseUrl}/${id}`;

    dispatch(updatingProduct());

    const response = await axios({
      url: url,
      method: 'put',
      data: data,
    }).catch((error) => {
      dispatch(updateProductFailure(error.message));
    });

    if (response) {
      const product = response.data;
      product.categories = getIds(product.categories);
      product.tags = getIds(product.tags);
      dispatch(updateProductSuccess(product));
    }
  };
};

export const deleteProduct = (id) => {
  return async (dispatch, getState) => {
    let url = `${baseUrl}/${id}`;

    dispatch(deletingProduct());

    const response = await axios({
      url: url,
      method: 'delete',
    }).catch((error) => {
      dispatch(deleteProductFailure(error.message));
    });

    if (response) {
      dispatch(deleteProductSuccess(id));
    }
  };
};

const loadingProducts = () => {
  return {
    type: LOADING_PRODUCTS,
  };
};

const setProductsListTotal = (total) => {
  return {
    type: SET_PRODUCTS_LIST_TOTAL,
    payload: total,
  };
};

const addListRequest = (req) => {
  return {
    type: ADD_PRODUCTS_LIST_REQUEST,
    payload: req,
  };
};

const setListCurrentPage = (ids) => {
  return {
    type: SET_PRODUCTS_LIST_CURRENT_PAGE,
    payload: ids,
  };
};

export const loadProductsSuccess = (products) => {
  return {
    type: LOAD_PRODUCTS_SUCCESS,
    payload: {
      items: buildObjectOfItems(deleteKeys(products, ['currency', 'product_type'])),
    },
  };
};

const loadProductsFailure = (message) => {
  return {
    type: LOAD_PRODUCTS_FAILURE,
    payload: message,
  };
};

const creatingProduct = () => {
  return {
    type: CREATING_PRODUCT,
  };
};

const createProductSuccess = (product) => {
  return {
    type: CREATE_PRODUCT_SUCCESS,
    payload: product,
  };
};

const createProductFailure = (message) => {
  return {
    type: CREATE_PRODUCT_FAILURE,
    payload: message,
  };
};

const loadingProductDetails = () => {
  return {
    type: LOADING_PRODUCT_DETAILS,
  };
};

const getProductDetailsSuccess = (product) => {
  return {
    type: GET_PRODUCT_DETAILS_SUCCESS,
    payload: product,
  };
};

const getProductDetailsFailure = (message) => {
  return {
    type: GET_PRODUCT_DETAILS_FAILURE,
    payload: message,
  };
};

const updatingProduct = () => {
  return {
    type: UPDATING_PRODUCT,
  };
};

const updateProductSuccess = (product) => {
  return {
    type: UPDATE_PRODUCT_SUCCESS,
    payload: deleteKeys([product], ['currency', 'product_type'])[0],
  };
};

const updateProductFailure = (message) => {
  return {
    type: UPDATE_PRODUCT_FAILURE,
    payload: message,
  };
};

const deletingProduct = () => {
  return {
    type: DELETING_PRODUCT,
  };
};

const deleteProductSuccess = (id) => {
  return {
    type: DELETE_PRODUCT_SUCCESS,
    payload: id,
  };
};

const deleteProductFailure = (message) => {
  return {
    type: DELETE_PRODUCT_FAILURE,
    payload: message,
  };
};
