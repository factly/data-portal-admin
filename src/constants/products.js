export const LOADING_PRODUCTS = 'LOADING_PRODUCTS';
export const LOAD_PRODUCTS_SUCCESS = 'LOAD_PRODUCTS_SUCCESS';
export const ADD_PRODUCTS_LIST_REQUEST = 'ADD_PRODUCTS_LIST_REQUEST';
export const SET_PRODUCTS_LIST_CURRENT_PAGE = 'SET_PRODUCTS_LIST_CURRENT_PAGE';
export const SET_PRODUCTS_LIST_TOTAL = 'SET_PRODUCTS_LIST_TOTAL';
export const LOAD_PRODUCTS_FAILURE = 'LOAD_PRODUCTS_FAILURE';
export const CREATING_PRODUCT = 'CREATING_PRODUCT';
export const CREATE_PRODUCT_SUCCESS = 'CREATE_PRODUCT_SUCCESS';
export const CREATE_PRODUCT_FAILURE = 'CREATE_PRODUCT_FAILURE';
export const LOADING_PRODUCT_DETAILS = 'LOADING_PRODUCT_DETAILS';
export const GET_PRODUCT_DETAILS_SUCCESS = 'GET_PRODUCT_DETAILS_SUCCESS';
export const GET_PRODUCT_DETAILS_FAILURE = 'GET_PRODUCT_DETAILS_FAILURE';
export const UPDATING_PRODUCT = 'UPDATING_PRODUCT';
export const UPDATE_PRODUCT_SUCCESS = 'UPDATE_PRODUCT_SUCCESS';
export const UPDATE_PRODUCT_FAILURE = 'UPDATE_PRODUCT_FAILURE';
export const DELETING_PRODUCT = 'DELETING_PRODUCT';
export const DELETE_PRODUCT_SUCCESS = 'DELETE_PRODUCT_SUCCESS';
export const DELETE_PRODUCT_FAILURE = 'DELETE_PRODUCT_FAILURE';

export const baseUrl = `${process.env.REACT_APP_API_URL}/products`;
