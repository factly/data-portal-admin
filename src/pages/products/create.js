import React, { useState } from 'react';
import { Form, Input, Button, notification, Select } from 'antd';

const { Option } = Select;

const formItemLayout = {
  labelCol: { span: 6 },
  wrapperCol: { span: 14 },
};

const ProductCreate = (props) => {
  const [tags, setTags] = useState([]);
  const [categories, setCategories] = useState([]);
  const [currencies, setCurrencies] = useState([]);
  const [productType, setProductType] = useState([]);

  React.useEffect(() => {
    fetch(process.env.REACT_APP_API_URL + '/tags')
      .then((data) => data.json())
      .then((data) => {
        setTags(data.nodes);
      });
    fetch(process.env.REACT_APP_API_URL + '/categories')
      .then((data) => data.json())
      .then((data) => {
        setCategories(data.nodes);
      });
    fetch(process.env.REACT_APP_API_URL + '/currencies')
      .then((data) => data.json())
      .then((data) => {
        setCurrencies(data.nodes);
      });
    fetch(process.env.REACT_APP_API_URL + '/types')
      .then((data) => data.json())
      .then((data) => {
        setProductType(data.nodes);
      });
  }, []);

  function handleChange(value) {
    console.log(`selected ${value}`);
  }

  const onFinish = (values) => {
    values.currency_id = parseInt(values.currency_id);
    values.product_type_id = parseInt(values.product_type_id);
    values.price = parseInt(values.price);
    values.category_ids = values.category_ids.map((c) => parseInt(c));
    values.tag_ids = values.tag_ids.map((t) => parseInt(t));
    fetch(process.env.REACT_APP_API_URL + '/products', {
      method: 'POST',
      body: JSON.stringify(values),
    })
      .then((res) => {
        if (res.status === 201) {
          return res.json();
        } else {
          throw new Error(res.status);
        }
      })
      .then((_) => {
        notification.success({
          message: 'Success',
          description: 'Product succesfully added',
        });
        props.history.push('/products');
      })
      .catch((res) => {
        notification.error({
          message: 'Error',
          description: 'Something went wrong',
        });
      });
  };

  return (
    <Form name="products_create" {...formItemLayout} onFinish={onFinish}>
      <Form.Item
        label="Title"
        name="title"
        rules={[
          {
            required: true,
            message: 'Please enter title!',
          },
        ]}
      >
        <Input placeholder="Ex. Crime In India" />
      </Form.Item>

      <Form.Item
        label="Slug"
        name="slug"
        rules={[
          {
            required: true,
            message: 'Please enter slug!',
          },
        ]}
      >
        <Input placeholder="Ex. crime-in-india" />
      </Form.Item>

      <Form.Item
        label="Currency"
        name="currency_id"
        rules={[
          {
            required: true,
            message: 'Please enter currency!',
          },
        ]}
      >
        <Select
          showSearch
          style={{ width: '100%' }}
          placeholder="Select currency"
          onChange={handleChange}
          filterOption={(input, option) =>
            option.props.children.toLowerCase().indexOf(input.toLowerCase()) >= 0
          }
        >
          {currencies.length > 0
            ? currencies.map((c) => <Option key={c.id}>{c.iso_code}</Option>)
            : []}
        </Select>
      </Form.Item>

      <Form.Item
        label="Price"
        name="price"
        rules={[
          {
            required: true,
            message: 'Please enter price!',
          },
        ]}
      >
        <Input placeholder="Ex. 1999" />
      </Form.Item>

      <Form.Item
        label="Category"
        name="category_ids"
        rules={[
          {
            required: true,
            message: 'Please select atleast one category!',
          },
        ]}
      >
        <Select
          mode="tags"
          style={{ width: '100%' }}
          placeholder="Select categories"
          onChange={handleChange}
          filterOption={(input, option) =>
            option.props.children.toLowerCase().indexOf(input.toLowerCase()) >= 0
          }
        >
          {categories.length > 0 ? categories.map((c) => <Option key={c.id}>{c.slug}</Option>) : []}
        </Select>
      </Form.Item>

      <Form.Item
        label="Tag"
        name="tag_ids"
        rules={[
          {
            required: true,
            message: 'Please select atleast one tag!',
          },
        ]}
      >
        <Select
          mode="tags"
          style={{ width: '100%' }}
          placeholder="Select tags"
          onChange={handleChange}
          filterOption={(input, option) =>
            option.props.children.toLowerCase().indexOf(input.toLowerCase()) >= 0
          }
        >
          {tags.length > 0 ? tags.map((t) => <Option key={t.id}>{t.slug}</Option>) : []}
        </Select>
      </Form.Item>

      <Form.Item
        label="Status"
        name="status"
        rules={[
          {
            required: true,
            message: 'Please enter status!',
          },
        ]}
      >
        <Select
          showSearch
          style={{ width: '100%' }}
          placeholder="Select status"
          defaultValue={'Hide'}
          onChange={handleChange}
          filterOption={(input, option) =>
            option.props.children.toLowerCase().indexOf(input.toLowerCase()) >= 0
          }
        >
          <Option key={'Hide'}>Hide</Option>
          <Option key={'Unhide'}>Unhide</Option>
        </Select>
      </Form.Item>

      <Form.Item
        label="Type"
        name="product_type_id"
        rules={[
          {
            required: true,
            message: 'Please enter type!',
          },
        ]}
      >
        <Select
          showSearch
          style={{ width: '100%' }}
          placeholder="Select type"
          onChange={handleChange}
          filterOption={(input, option) =>
            option.props.children.toLowerCase().indexOf(input.toLowerCase()) >= 0
          }
        >
          {productType.length > 0
            ? productType.map((p) => <Option key={p.id}>{p.name}</Option>)
            : []}
        </Select>
      </Form.Item>

      <Form.Item wrapperCol={{ span: 12, offset: 6 }}>
        <Button type="primary" htmlType="submit">
          Submit
        </Button>
      </Form.Item>
    </Form>
  );
};

export default ProductCreate;
