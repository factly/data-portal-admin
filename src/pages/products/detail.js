import React from 'react';
import { useParams } from 'react-router-dom';
import { useSelector } from 'react-redux';

import ProductDetails from './components/ProductDetails';

const ProductDetail = () => {
  const { id } = useParams();
  const { product, currencies, tags } = useSelector(({ products, currencies, tags }) => ({
    currencies: currencies.items,
    tags: tags.items,
    product: products.items[id],
  }));

  const currencyCode = currencies[product.currency_id].iso_code;
  const productTags = [];
  product.tags.forEach((id) => productTags.push(tags[id]));

  return <ProductDetails product={product} productTags={productTags} currencyCode={currencyCode} />;
};

export default ProductDetail;
