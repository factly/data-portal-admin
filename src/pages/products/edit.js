import React, { useState } from 'react';
import { Form, Input, Button, notification, Select } from 'antd';
import { useParams } from 'react-router-dom';

const { Option } = Select;

const formItemLayout = {
  labelCol: { span: 6 },
  wrapperCol: { span: 14 },
};

const ProductEdit = (props) => {
  const [tags, setTags] = useState([]);
  const [categories, setCategories] = useState([]);
  const [currencies, setCurrencies] = useState([]);
  const [productType, setProductType] = useState([]);
  const [product, setProduct] = useState();

  const { id } = useParams();
  React.useEffect(() => {
    fetch(process.env.REACT_APP_API_URL + `/products/${id}`)
      .then((data) => data.json())
      .then((data) => {
        setProduct(data);
      })
      .catch((err) => {
        notification.error({
          message: 'Error',
          description: 'Something went wrong',
        });
      });
    fetch(process.env.REACT_APP_API_URL + '/tags')
      .then((data) => data.json())
      .then((data) => {
        setTags(data.nodes);
      });
    fetch(process.env.REACT_APP_API_URL + '/categories')
      .then((data) => data.json())
      .then((data) => {
        setCategories(data.nodes);
      });
    fetch(process.env.REACT_APP_API_URL + '/currencies')
      .then((data) => data.json())
      .then((data) => {
        setCurrencies(data.nodes);
      });
    fetch(process.env.REACT_APP_API_URL + '/types')
      .then((data) => data.json())
      .then((data) => {
        setProductType(data.nodes);
      });
  }, [id]);

  function handleChange(value) {
    console.log(`selected ${value}`);
  }

  const onFinish = (values) => {
    values.currency_id = parseInt(values.currency_id);
    values.product_type_id = parseInt(values.product_type_id);
    values.price = parseInt(values.price);
    values.category_ids = values.category_ids.map((id) => parseInt(id));
    values.tag_ids = values.tag_ids.map((id) => parseInt(id));

    fetch(process.env.REACT_APP_API_URL + `/products/${id}`, {
      method: 'PUT',
      body: JSON.stringify(values),
    })
      .then((res) => {
        if (res.ok) {
          return res.json();
        } else {
          throw new Error(res.status);
        }
      })
      .then((_) => {
        notification.success({
          message: 'Success',
          description: 'Product succesfully updated',
        });
        props.history.push('/products');
      })
      .catch((err) => {
        notification.error({
          message: 'Error',
          description: 'Something went wrong',
        });
      });
  };

  const initialValues = {
    slug: product ? product.slug : '',
    title: product ? product.title : '',
    price: product ? product.price : '',
    currency_id: product ? product.Currency.id : '',
    category_ids: product ? product.categories.map((category) => category.id) : [],
    tag_ids: product ? product.tags.map((tag) => tag.id) : [],
    status: product ? product.status : '',
    product_type_id: product ? product.ProductType.id : '',
  };

  return !product ? null : (
    <Form
      name="products_update"
      {...formItemLayout}
      initialValues={initialValues}
      onFinish={onFinish}
    >
      <Form.Item
        label="Title"
        name="title"
        rules={[
          {
            required: true,
            message: 'Please enter title!',
          },
        ]}
      >
        <Input placeholder="Ex. Crime In India" />
      </Form.Item>

      <Form.Item
        label="Slug"
        name="slug"
        rules={[
          {
            required: true,
            message: 'Please enter slug!',
          },
        ]}
      >
        <Input placeholder="Ex. crime-in-india" />
      </Form.Item>

      <Form.Item
        label="Currency"
        name="currency_id"
        rules={[
          {
            required: true,
            message: 'Please enter currency!',
          },
        ]}
      >
        <Select
          showSearch
          style={{ width: '100%' }}
          placeholder="Select currency"
          onChange={handleChange}
          filterOption={(input, option) =>
            option.props.children.toLowerCase().indexOf(input.toLowerCase()) >= 0
          }
        >
          {currencies.length > 0
            ? currencies.map((currency) => (
                <Select.Option key={currency.id} value={currency.id}>
                  {currency.name}
                </Select.Option>
              ))
            : []}
        </Select>
      </Form.Item>

      <Form.Item
        label="Price"
        name="price"
        rules={[
          {
            required: true,
            message: 'Please enter price!',
          },
        ]}
      >
        <Input placeholder="Ex. 1999" />
      </Form.Item>

      <Form.Item
        label="Category"
        name="category_ids"
        rules={[
          {
            required: true,
            message: 'Please select atleast one category!',
          },
        ]}
      >
        <Select
          mode="multiple"
          style={{ width: '100%' }}
          placeholder="Select categories"
          onChange={handleChange}
          filterOption={(input, option) =>
            option.props.children.toLowerCase().indexOf(input.toLowerCase()) >= 0
          }
        >
          {categories.length > 0
            ? categories.map((category) => (
                <Select.Option key={category.id} value={category.id}>
                  {category.title}
                </Select.Option>
              ))
            : []}
        </Select>
      </Form.Item>

      <Form.Item
        label="Tag"
        name="tag_ids"
        rules={[
          {
            required: true,
            message: 'Please select atleast one tag!',
          },
        ]}
      >
        <Select
          mode="multiple"
          style={{ width: '100%' }}
          placeholder="Select tags"
          onChange={handleChange}
          filterOption={(input, option) =>
            option.props.children.toLowerCase().indexOf(input.toLowerCase()) >= 0
          }
        >
          {tags.length > 0
            ? tags.map((tag) => (
                <Select.Option key={tag.id} value={tag.id}>
                  {tag.title}
                </Select.Option>
              ))
            : []}
        </Select>
      </Form.Item>

      <Form.Item
        label="Status"
        name="status"
        rules={[
          {
            required: true,
            message: 'Please enter status!',
          },
        ]}
      >
        <Select
          showSearch
          style={{ width: '100%' }}
          placeholder="Select status"
          onChange={handleChange}
          filterOption={(input, option) =>
            option.props.children.toLowerCase().indexOf(input.toLowerCase()) >= 0
          }
        >
          <Option key={1} value={'Show'}>
            Show
          </Option>
          <Option key={2} value={'Hide'}>
            Hide
          </Option>
        </Select>
      </Form.Item>

      <Form.Item
        label="Type"
        name="product_type_id"
        rules={[
          {
            required: true,
            message: 'Please enter type!',
          },
        ]}
      >
        <Select
          showSearch
          style={{ width: '100%' }}
          placeholder="Select type"
          onChange={handleChange}
          filterOption={(input, option) =>
            option.props.children.toLowerCase().indexOf(input.toLowerCase()) >= 0
          }
        >
          {productType.length > 0
            ? productType.map((product) => (
                <Select.Option key={product.id} value={product.id}>
                  {product.name}
                </Select.Option>
              ))
            : []}
        </Select>
      </Form.Item>

      <Form.Item wrapperCol={{ span: 12, offset: 6 }}>
        <Button type="primary" htmlType="submit">
          Submit
        </Button>
      </Form.Item>
    </Form>
  );
};

export default ProductEdit;
