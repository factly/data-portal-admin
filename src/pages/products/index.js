import React, { useState } from 'react';
import { Table, Input, Button, Popconfirm, Form, notification } from 'antd';
import moment from 'moment';
import { Link } from 'react-router-dom';
import { EditOutlined, DeleteOutlined } from '@ant-design/icons';

const EditableCell = ({ editing, dataIndex, title, record, index, children, ...restProps }) => {
  return (
    <td {...restProps}>
      {editing ? (
        <Form.Item
          name={dataIndex}
          style={{
            margin: 0,
          }}
          rules={[
            {
              required: true,
              message: `Please enter ${title}!`,
            },
          ]}
        >
          <Input />
        </Form.Item>
      ) : (
        children
      )}
    </td>
  );
};

const Products = (props) => {
  const [form] = Form.useForm();
  const [data, setData] = useState([]);
  const [total, setTotal] = useState(0);
  const [editingKey, setEditingKey] = useState('');

  React.useEffect(() => {
    fetch(process.env.REACT_APP_API_URL + '/products')
      .then((data) => data.json())
      .then((data) => {
        setData(data.nodes);
        setTotal(data.total);
      });
  }, []);

  const get = (page, limit) => {
    cancel();
    fetch(process.env.REACT_APP_API_URL + '/products?page=' + page + '&limit=' + limit)
      .then((data) => data.json())
      .then((data) => {
        setData(data.nodes);
        setTotal(data.total);
      });
  };

  const isEditing = (record) => record.id === editingKey;

  const cancel = () => {
    setEditingKey('');
  };

  const deleteProduct = (key) => {
    const index = data.findIndex((item) => item.id === key);
    if (index > -1) {
      fetch(process.env.REACT_APP_API_URL + '/products/' + key, {
        method: 'DELETE',
      })
        .then((res) => {
          if (res.status === 200) {
            const newData = [...data];
            newData.splice(index, 1);
            setData(newData);
            notification.success({
              message: 'Success',
              description: 'Product succesfully deleted',
            });
          }
        })
        .catch((err) => {
          notification.error({
            message: 'Error',
            description: 'Something went wrong',
          });
        });
    }
  };

  const columns = [
    {
      title: 'Title',
      dataIndex: 'title',
      width: '10%',
    },
    {
      title: 'Slug',
      dataIndex: 'slug',
      width: '10%',
    },
    {
      title: 'Currency',
      render: (record) => record.currency.iso_code,
      width: '10%',
    },
    {
      title: 'Price',
      dataIndex: 'price',
      width: '10%',
    },
    {
      title: 'Status',
      render: (record) => record.status,
      width: '10%',
    },
    {
      title: 'Categories',
      render: (record) => record.categories.map((c) => c.title).join(', '),
      width: '20%',
    },
    {
      title: 'Tags',
      render: (record) => record.tags.map((t) => t.title).join(', '),
      width: '20%',
    },
    {
      title: 'Created At',
      dataIndex: 'CreatedAt',
      width: '10%',
      render: (_, record) => {
        return <span title={record.CreatedAt}>{moment(record.CreatedAt).fromNow()}</span>;
      },
    },
    {
      title: 'Operation',
      dataIndex: 'operation',
      render: (_, record) => {
        return (
          <span>
            <Button
              type="primary"
              icon={<EditOutlined />}
              disabled={editingKey !== ''}
              onClick={() => {
                props.history.push(`/products/detail/${record.id}`);
              }}
              style={{
                marginRight: 8,
              }}
            >
              Edit
            </Button>
            <Popconfirm
              disabled={editingKey !== ''}
              title="Sure to delete?"
              onConfirm={() => deleteProduct(record.id)}
            >
              <Button disabled={editingKey !== ''} icon={<DeleteOutlined />}>
                Delete
              </Button>
            </Popconfirm>
          </span>
        );
      },
    },
  ];

  const mergedColumns = columns.map((col) => {
    if (!col.editable) {
      return col;
    }

    return {
      ...col,
      onCell: (record) => ({
        record,
        dataIndex: col.dataIndex,
        title: col.title,
        editing: isEditing(record),
      }),
    };
  });

  return (
    <div>
      <Link to={'/products/create'}>
        <Button type="primary" style={{ marginBottom: 16 }}>
          Add Product
        </Button>
      </Link>
      <Form form={form} component={false}>
        <Table
          components={{
            body: {
              cell: EditableCell,
            },
          }}
          bordered
          rowKey="id"
          dataSource={data}
          columns={mergedColumns}
          rowClassName="editable-row"
          pagination={{
            defaultPageSize: 5,
            onChange: get,
            total: total,
          }}
        />
      </Form>
    </div>
  );
};

export default Products;
