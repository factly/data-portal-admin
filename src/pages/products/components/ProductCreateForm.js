import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory } from 'react-router-dom';

import { Form, Input, Button, notification, Select } from 'antd';

import { loadTags } from '../../../actions/tags';
import { loadCurrencies } from '../../../actions/currencies';

const { Option } = Select;

const formItemLayout = {
  labelCol: { span: 6 },
  wrapperCol: { span: 14 },
};

const ProductCreateForm = ({ onSubmit, data = {} }) => {
  const history = useHistory();
  const dispatch = useDispatch();
  const { tags, currencies } = useSelector(({ tags, currencies }) => {
    return {
      tags: Object.values(tags.items),
      currencies: Object.values(currencies.items),
    };
  });

  React.useEffect(() => {
    dispatch(loadTags());
    dispatch(loadCurrencies());
  }, []);

  const onFinish = (values) => {
    values.currency_id = parseInt(values.currency_id);
    values.price = parseInt(values.price);
    values.tag_ids = values.tag_ids.map((id) => parseInt(id));

    onSubmit(values)
      .then(() => {
        notification.success({
          message: 'Success',
          description: 'Product added succesfully',
        });
        history.push('/products');
      })
      .catch(() => {
        notification.error({
          message: 'Error',
          description: 'Something went wrong',
        });
      });
  };

  return (
    <Form name="products_create" initialValues={data} {...formItemLayout} onFinish={onFinish}>
      <Form.Item
        label="Title"
        name="title"
        rules={[
          {
            required: true,
            message: 'Please enter title!',
          },
        ]}
      >
        <Input placeholder="Ex. Crime In India" />
      </Form.Item>

      <Form.Item
        label="Slug"
        name="slug"
        rules={[
          {
            required: true,
            message: 'Please enter slug!',
          },
        ]}
      >
        <Input placeholder="Ex. crime-in-india" />
      </Form.Item>

      <Form.Item
        label="Currency"
        name="currency_id"
        rules={[
          {
            required: true,
            message: 'Please enter currency!',
          },
        ]}
      >
        <Select
          showSearch
          style={{ width: '100%' }}
          placeholder="Select currency"
          filterOption={(input, option) =>
            option.props.children.toLowerCase().indexOf(input.toLowerCase()) >= 0
          }
        >
          {currencies.length > 0
            ? currencies.map((currency) => (
                <Option key={currency.id} value={currency.id}>
                  {currency.iso_code}
                </Option>
              ))
            : []}
        </Select>
      </Form.Item>

      <Form.Item
        label="Price"
        name="price"
        rules={[
          {
            required: true,
            message: 'Please enter price!',
          },
        ]}
      >
        <Input type="number" placeholder="Ex. 1999" />
      </Form.Item>

      <Form.Item
        label="Tag"
        name="tag_ids"
        rules={[
          {
            required: true,
            message: 'Please select atleast one tag!',
          },
        ]}
      >
        <Select
          mode="multiple"
          style={{ width: '100%' }}
          placeholder="Select tags"
          filterOption={(input, option) =>
            option.props.children.toLowerCase().indexOf(input.toLowerCase()) >= 0
          }
        >
          {tags.length > 0
            ? tags.map((tag) => (
                <Option key={tag.id} value={tag.id}>
                  {tag.slug}
                </Option>
              ))
            : []}
        </Select>
      </Form.Item>

      <Form.Item
        label="Status"
        name="status"
        rules={[
          {
            required: true,
            message: 'Please enter status!',
          },
        ]}
      >
        <Select
          showSearch
          style={{ width: '100%' }}
          placeholder="Select status"
          filterOption={(input, option) =>
            option.props.children.toLowerCase().indexOf(input.toLowerCase()) >= 0
          }
        >
          <Option key={1} value={'Show'}>
            Show
          </Option>
          <Option key={2} value={'Hide'}>
            Hide
          </Option>
        </Select>
      </Form.Item>

      <Form.Item wrapperCol={{ span: 12, offset: 6 }}>
        <Button type="primary" htmlType="submit">
          Submit
        </Button>
      </Form.Item>
    </Form>
  );
};

export default ProductCreateForm;
