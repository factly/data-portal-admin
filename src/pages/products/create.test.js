import React from 'react';
import { useHistory } from 'react-router-dom';
import { useDispatch, Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount, shallow } from 'enzyme';
import { act } from '@testing-library/react';

import '../../matchMedia.mock';
import CreateProduct from './create';
import * as actions from '../../actions/products';
import ProductCreateForm from './components/ProductCreateForm';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn(),
}));

jest.mock('../../actions/products', () => ({
  createProduct: jest.fn(),
}));

describe('Products create component', () => {
  let store;
  let mockedDispatch;

  store = mockStore({
    products: {
      loading: false,
      ids: [],
      req: [],
      items: {},
      total: 0,
    },
    currencies: {
      loading: false,
      ids: [],
      req: [],
      items: {},
      total: 0,
    },
    tags: {
      loading: false,
      ids: [],
      req: [],
      items: {},
      total: 0,
    },
    datasets: {
      loading: false,
      ids: [],
      req: [],
      items: {},
      total: 0,
    },
  });
  store.dispatch = jest.fn(() => ({}));
  mockedDispatch = jest.fn(() => Promise.resolve({}));
  useDispatch.mockReturnValue(mockedDispatch);
  describe('snapshot testing', () => {
    it('should render the component', () => {
      const component = shallow(<CreateProduct />);
      expect(component).toMatchSnapshot();
    });
  });
  describe('component testing', () => {
    let wrapper;
    afterEach(() => {
      wrapper.unmount();
    });
    it('should call createProduct', (done) => {
      actions.createProduct.mockReset();
      const push = jest.fn();
      useHistory.mockReturnValueOnce({ push });
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <CreateProduct />
          </Provider>,
        );
      });
      wrapper.find(ProductCreateForm).props().onSubmit({ test: 'test' });
      setTimeout(() => {
        expect(actions.createProduct).toHaveBeenCalledWith({ test: 'test' });
        done();
      }, 0);
    });
  });
});
