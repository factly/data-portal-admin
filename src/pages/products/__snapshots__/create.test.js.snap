// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products create component snapshot testing should render the component 1`] = `
<ProductCreateForm
  onSubmit={[Function]}
/>
`;
